re.c

```diff
@@ -2348,6 +2348,82 @@ match_named_captures(VALUE match)
     return hash;
 }
 
+/*
+ *  call-seq:
+ *    deconstruct_keys(array_of_names) -> hash
+ *
+ *  Returns a hash of the named captures for the given names.
+ *
+ *    m = /(?<hours>\d{2}):(?<minutes>\d{2}):(?<seconds>\d{2})/.match("18:37:22")
+ *    m.deconstruct_keys([:hours, :minutes]) # => {:hours => "18", :minutes => "37"}
+ *    m.deconstruct_keys(nil) # => {:hours => "18", :minutes => "37", :seconds => "22"}
+ *
+ *  Returns an empty hash of no named captures were defined:
+ *
+ *    m = /(\d{2}):(\d{2}):(\d{2})/.match("18:37:22")
+ *    m.deconstruct_keys(nil) # => {}
+ *
+ */
+static VALUE
+match_deconstruct_keys(VALUE match, VALUE keys)
+{
+    VALUE h, names;
+    long i;
+
+    match_check(match);
+
+    if (NIL_P(RMATCH(match)->regexp)) {
+        return rb_hash_new_with_size(0);
+    }
+
+    names = rb_reg_names(RMATCH(match)->regexp);
+
+    if (NIL_P(keys)) {
+        keys = rb_ary_new2(RARRAY_LEN(names));
+
+        for (i=0; i<RARRAY_LEN(names); i++) {
+            rb_ary_push(keys, rb_str_intern(RARRAY_AREF(names, i)));
+        }
+    }
+
+    if (UNLIKELY(!RB_TYPE_P(keys, T_ARRAY))) {
+        rb_raise(rb_eTypeError,
+                 "wrong argument type %"PRIsVALUE" (expected Array or nil)",
+                 rb_obj_class(keys));
+
+    }
+
+    if (RARRAY_LEN(names) < RARRAY_LEN(keys)) {
+        return rb_hash_new_with_size(0);
+    }
+
+    h = rb_hash_new_with_size(RARRAY_LEN(keys));
+
+    for (i=0; i<RARRAY_LEN(keys); i++) {
+        VALUE key = RARRAY_AREF(keys, i);
+        VALUE name;
+
+        if (UNLIKELY(!SYMBOL_P(key))) {
+            rb_raise(rb_eTypeError,
+                 "wrong argument type %"PRIsVALUE" (expected Symbol)",
+                 rb_obj_class(key));
+        }
```

Comment:
@k-tsj Not sure if we need to be defensive here; what do you think?

Reply:
It is needed because `rb_sym2str` requires an argument is a symbol.